custom_metrics/wpt_bodies.js

```diff
@@ -0,0 +1,1366 @@
+            let items = d.querySelectorAll('.'+name);
+            if (items.length > 0) {
+              target.microformats2_types.push({name: name, count: items.length});
+              target.items_by_format.microformats2 += items.length;
+            }
+          });
+
+          return target;
+        }
+
+        let result = {};
+        result.rendered = gatherStructuredData(document);
+
+        let rawHtmlDocument = getRawHtmlDocument();
+
+        if (rawHtmlDocument) {
+          result.raw = gatherStructuredData(rawHtmlDocument);
+        }
+        
+        return result; 
+      }
+      catch(e) {
+        return logError("structured_data", e);
+      }
+    })(),
+
+    // data from the original html
+    // Used by: SEO, Markup  
+    // Backup: queiries on the body - expensive
+    'raw_html': (() => {  
+      try {
+        let result = {};
+
+        let rawHtml = getRawHtml();
+
+        if (rawHtml) {
+
+          result.body = !!rawHtml.match(/<body/g);
+          result.html = !!rawHtml.match(/<html/g);
+          result.head = !!rawHtml.match(/<head/g);
+          result.size = rawHtml.length;
+
+          let headmatch = rawHtml.match(/<head.*<\/head>/gsi); // s = match newlines
+
+          if (headmatch) {
+            result.head_size = headmatch[0].length;
+          }
+
+          let bodymatch = rawHtml.match(/<body.*<\/body>/gsi); // s = match newlines
+
+          if (bodymatch) {
+            result.body_size = bodymatch[0].length;
+          }
+
+          let commentMatches = rawHtml.match(/<!--/g);
+
+          if (commentMatches) {
+            result.comment_count = commentMatches.length;
+          }
+
+          let ifCommentMatches = rawHtml.match(/<!-- *\[ *if/gsi);
+
+          if (ifCommentMatches) {
+            result.conditional_comment_count = ifCommentMatches.length;
+          }
+        }
+
+        return result;
+      }
+      catch(e) {
+        return logError("raw_html", e);
+      }
+    })(),
+    // canonicals used in the page and http header
+    // Used by: SEO
+    // Backup: link nodes in almanac.js
+    'canonicals': (() => {
+      try {
+        function processCanonical(c) {
+          if (c === result.url) 
+            result.self_canonical = true;
+        else 
+            result.other_canonical = true;
+
+          if (result.canonicals[c]) 
+            result.canonicals[c]++;
+          else 
+            result.canonicals[c] = 1;
+        }
+
+        let result = {rendered: {}, raw: {}, self_canonical: false, other_canonical: false, canonicals: {}};
+        result.url = document.location.href.split("#")[0];
+
+        // headers
+        let canonicalLinkHeaders = [];
+          // Link: <https://example.com/page-b>; rel="canonical"
+        let linkHeaders = getResponseHeaders("link");
+        if (linkHeaders) {
+          linkHeaders.forEach((h) => {
+            let matches = h.matchAll(/<([^>]*)> *; *rel=['"]?canonical['"]?/gi);
```

Comment:
Should these space characters (` *`) be explicit `\s*`? And do you want to ensure at least one non-`>` character?
```suggestion
            let matches = h.matchAll(/<([^>]+)>\s*;\s*rel=['"]?canonical['"]?/gi);
```